examples/fileserver/fileserver.go

```diff
@@ -25,6 +29,89 @@ var (
 	vhost              = flag.Bool("vhost", false, "Enables virtual hosting by prepending the requested path with the requested hostname")
 )
 
+func hellloHandler(ctx *fasthttp.RequestCtx) {
+	fmt.Fprintf(ctx, "<h1>Hello, world!</h1>")
+
+	ctx.SetContentType("text/html; charset=utf8")
+
+	// Set arbitrary headers
+	ctx.Response.Header.Set("X-My-Header", "my-header-value")
+
+	// Set cookies
+	var c fasthttp.Cookie
+	c.SetKey("cookie-name")
+	c.SetValue("cookie-value")
+	ctx.Response.Header.SetCookie(&c)
+}
+
+func jsonHandler(ctx *fasthttp.RequestCtx) {
+	type Vertex struct {
+		X, Y int
+		Rock string
+		Z    int
+	}
+	v1 := Vertex{1, 2, "Thay Cuong is happy", 10}
+	b, err := json.Marshal(&v1)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+
+	ctx.Response.Header.Add("Content-Encoding", "gzip")
+	ctx.SetContentType("application/json; charset=utf8")
+	/* This is ok
+	ctx.Write(
+		fasthttp.AppendGzipBytes(
+			nil, b),
+	)*/
+	fasthttp.WriteGzip(ctx, b)
+}
+
+
+//Demo streaming
+func streamHandler(ctx *fasthttp.RequestCtx) {
+	type Geolocation struct {
+		Altitude  float64
+		Latitude  float64
+		Longitude float64
+	}
+
+	locations := []Geolocation{
+		{-97, 37.819929, -122.478255},
+		{1899, 39.096849, -120.032351},
+		{2619, 37.865101, -119.538329},
+		{42, 33.812092, -117.918974},
+		{15, 37.77493, -122.419416},
+		{2613, 67.865101, -119.538329},
+		{44, 53.812092, -117.918974},
+		{25, 57.77493, -122.419416},
+	}
+
+	ctx.SetContentType("application/json; charset=utf8")
+
+	closeStream := make(chan int)
+	streamReader := fasthttp.NewStreamReader(func(w *bufio.Writer) {
+		for _, l := range locations {
+			if err := json.NewEncoder(w).Encode(l); err != nil {
+				continue
+			}
+			w.Flush()
+			time.Sleep(1 * time.Second)
+		}
+		closeStream <- 0  //Signal to close stream reader
```

Comment:
I have used https://godoc.org/github.com/erikdubbelboer/fasthttp#example-RequestCtx-SetBodyStreamWriter but I found that its behavior is not streaming. The server hold each trunk response until the last write.

In my example, the server responses trunk by trunk.

Reply:
Your question: Isn't it possible to just call streamReader.Close() here instead of using the chan?

My answer: We have to wait the last write of stream operation then call streamReader.Close(). We cannot call  streamReader.Close() too early

Reply:
But the result is basically the same. Now you have a goroutine waiting for the channel to be closed and then it immediately calls `streamReader.Close()`. So if you just call `streamReader.Close()` instead of closing the channel the difference is only a couple of CPU cycles for the closing of the channel.